File: src/humanoid_navigation2/humanoid_navigation2/fastlio_open3d_axis_adapter.py
#!/usr/bin/env python3
"""
fastlio_open3d_axis_adapter.py

这个节点专门解决本系统的 Fast-LIO 原始坐标轴问题。

背景：
  Fast-LIO 原始 camera_init/body 坐标轴为：
    x 朝左、y 朝下、z 朝后

  Nav2 / 2D map / grounded PCD 使用 ROS 标准导航坐标轴：
    x 朝前、y 朝左、z 朝上

如果让 open3d_loc 直接订阅 raw /odom 和 raw /fast_lio/cloud_registered，
它内部配准可以运行，但输出位姿会停留在 raw Fast-LIO 坐标约定里。
随后 prior_map_odom_bridge 再拿 TF 树里的标准 odom->base_footprint 去相乘，
就会把两套坐标轴混在一起，表现为方向错、平移错，严重时看起来像定位跳变。

本节点做两件事：
  1. 把 raw Fast-LIO 世界点云转换成标准轴 grounded 点云，供 open3d_loc 配准。
  2. 把 raw /odom 中的 camera_init->body 位姿转换成标准轴
     odom->prior_open3d_base，并发布同名 TF，供 bridge 精确计算 map->odom。

转换公式与 scripts/make_localization_pcd.py 保持一致：
  p_std = R_raw_to_std * (p_raw - t_initial_body_to_base_raw)

其中 t_initial_body_to_base_raw 来自 navigation2.launch.py 里的 body->base_footprint
平移。这样地图 PCD、实时点云、open3d 位姿输入都在同一套标准坐标里。
"""

import math
from typing import Tuple

import numpy as np
import rclpy
from geometry_msgs.msg import TransformStamped
from nav_msgs.msg import Odometry
from rclpy.node import Node
from rclpy.qos import HistoryPolicy, QoSProfile, ReliabilityPolicy
from sensor_msgs.msg import PointCloud2
import sensor_msgs_py.point_cloud2 as pc2
from tf2_ros import TransformBroadcaster


def normalize_quaternion(x: float, y: float, z: float, w: float) -> Tuple[float, float, float, float]:
    """归一化四元数，避免上游偶发的非单位四元数污染矩阵计算。"""
    norm = math.sqrt(x * x + y * y + z * z + w * w)
    if norm <= 1e-12:
        return 0.0, 0.0, 0.0, 1.0
    return x / norm, y / norm, z / norm, w / norm


def quaternion_to_matrix(x: float, y: float, z: float, w: float) -> np.ndarray:
    """ROS 四元数(x,y,z,w) -> 3x3 旋转矩阵。"""
    x, y, z, w = normalize_quaternion(x, y, z, w)
    return np.array(
        [
            [1.0 - 2.0 * (y * y + z * z), 2.0 * (x * y - w * z), 2.0 * (x * z + w * y)],
            [2.0 * (x * y + w * z), 1.0 - 2.0 * (x * x + z * z), 2.0 * (y * z - w * x)],
            [2.0 * (x * z - w * y), 2.0 * (y * z + w * x), 1.0 - 2.0 * (x * x + y * y)],
        ],
        dtype=np.float64,
    )


def matrix_to_quaternion(matrix: np.ndarray) -> Tuple[float, float, float, float]:
    """3x3 旋转矩阵 -> ROS 四元数(x,y,z,w)。"""
    m = matrix
    trace = float(np.trace(m))
    if trace > 0.0:
        s = math.sqrt(trace + 1.0) * 2.0
        qw = 0.25 * s
        qx = (m[2, 1] - m[1, 2]) / s
        qy = (m[0, 2] - m[2, 0]) / s
        qz = (m[1, 0] - m[0, 1]) / s
    elif m[0, 0] > m[1, 1] and m[0, 0] > m[2, 2]:
        s = math.sqrt(max(1.0 + m[0, 0] - m[1, 1] - m[2, 2], 1e-12)) * 2.0
        qw = (m[2, 1] - m[1, 2]) / s
        qx = 0.25 * s
        qy = (m[0, 1] + m[1, 0]) / s
        qz = (m[0, 2] + m[2, 0]) / s
    elif m[1, 1] > m[2, 2]:
        s = math.sqrt(max(1.0 + m[1, 1] - m[0, 0] - m[2, 2], 1e-12)) * 2.0
        qw = (m[0, 2] - m[2, 0]) / s
        qx = (m[0, 1] + m[1, 0]) / s
        qy = 0.25 * s
        qz = (m[1, 2] + m[2, 1]) / s
    else:
        s = math.sqrt(max(1.0 + m[2, 2] - m[0, 0] - m[1, 1], 1e-12)) * 2.0
        qw = (m[1, 0] - m[0, 1]) / s
        qx = (m[0, 2] + m[2, 0]) / s
        qy = (m[1, 2] + m[2, 1]) / s
        qz = 0.25 * s
    return normalize_quaternion(qx, qy, qz, qw)


class FastLioOpen3dAxisAdapter(Node):
    """把 Fast-LIO raw 轴输入转换成 open3d_loc 可直接使用的标准轴输入。"""

    def __init__(self):
        super().__init__("fastlio_open3d_axis_adapter")

        self.raw_odom_topic = self.declare_parameter("raw_odom_topic", "/odom").value
        self.raw_cloud_topic = self.declare_parameter("raw_cloud_topic", "/fast_lio/cloud_registered").value
        self.output_odom_topic = self.declare_parameter(
            "output_odom_topic", "/prior_localization/open3d_input_odom").value
        self.output_cloud_topic = self.declare_parameter(
            "output_cloud_topic", "/prior_localization/open3d_input_cloud").value
        self.odom_frame = self.declare_parameter("odom_frame", "odom").value
        self.output_base_frame = self.declare_parameter("output_base_frame", "prior_open3d_base").value
        self.publish_tf = bool(self.declare_parameter("publish_tf", True).value)

        # 与 make_localization_pcd.py 中的 R_CAMERA_TO_MAP 完全一致：
        #   X_std = -Z_raw, Y_std = X_raw, Z_std = -Y_raw
        self.raw_to_std = np.array(
            [
                [0.0, 0.0, -1.0],
                [1.0, 0.0, 0.0],
                [0.0, -1.0, 0.0],
            ],
            dtype=np.float64,
        )

        # body->base_footprint 的 raw 平移，来自 navigation2.launch.py。
        # 地图生成脚本会减掉这段平移，使 grounded PCD 的原点落在初始 base_footprint。
        offset = self.declare_parameter(
            "initial_body_to_base_translation_raw",
            [0.004, 1.215, 0.072],
        ).value
        self.initial_body_to_base_translation_raw = np.asarray(offset, dtype=np.float64)

        # body->base_footprint 的旋转。它把 raw body 轴旋到标准 base_footprint 轴。
        self.body_to_base_rotation_raw = np.array(
            [
                [0.0, 1.0, 0.0],
                [0.0, 0.0, -1.0],
                [-1.0, 0.0, 0.0],
            ],
            dtype=np.float64,
        )

        input_cloud_qos = QoSProfile(
            reliability=ReliabilityPolicy.BEST_EFFORT,
            history=HistoryPolicy.KEEP_LAST,
            depth=3,
        )
        output_cloud_qos = QoSProfile(
            reliability=ReliabilityPolicy.RELIABLE,
            history=HistoryPolicy.KEEP_LAST,
            depth=3,
        )

        self.odom_pub = self.create_publisher(Odometry, self.output_odom_topic, 30)
        # open3d_loc 的 C++ 订阅默认是 reliable。这里输出必须用 reliable，
        # 否则 DDS 会报 QoS 不兼容，外部定位节点收不到任何点云。
        self.cloud_pub = self.create_publisher(PointCloud2, self.output_cloud_topic, output_cloud_qos)
        self.tf_broadcaster = TransformBroadcaster(self)

        self.create_subscription(Odometry, self.raw_odom_topic, self.on_raw_odom, 50)
        self.create_subscription(PointCloud2, self.raw_cloud_topic, self.on_raw_cloud, input_cloud_qos)

        self.get_logger().info(
            "fastlio_open3d_axis_adapter started: "
            f"{self.raw_odom_topic}->{self.output_odom_topic}, "
            f"{self.raw_cloud_topic}->{self.output_cloud_topic}, "
            f"tf={self.odom_frame}->{self.output_base_frame}"
        )

    def on_raw_odom(self, msg: Odometry):
        """
        raw /odom 里通常是 camera_init->body。

        这里先算 raw camera_init 下的 base_footprint 位姿，再转到标准 odom：
          p_base_std = R_raw_to_std * (p_body_raw + R_body_raw * t_body_base - t_initial_base)
          R_base_std = R_raw_to_std * R_body_raw * R_body_to_base_raw

        初始时 p_body_raw=0、R_body_raw=I，所以 p_base_std=0、R_base_std=I。
        """
        pose = msg.pose.pose
        raw_rotation = quaternion_to_matrix(
            pose.orientation.x,
            pose.orientation.y,
            pose.orientation.z,
            pose.orientation.w,
        )
        raw_translation = np.array(
            [pose.position.x, pose.position.y, pose.position.z],
            dtype=np.float64,
        )

        base_translation_raw = raw_translation + raw_rotation @ self.initial_body_to_base_translation_raw
        base_translation_std = self.raw_to_std @ (
            base_translation_raw - self.initial_body_to_base_translation_raw
        )
        base_rotation_std = self.raw_to_std @ raw_rotation @ self.body_to_base_rotation_raw
        qx, qy, qz, qw = matrix_to_quaternion(base_rotation_std)

        out = Odometry()
        out.header.stamp = msg.header.stamp
        out.header.frame_id = self.odom_frame
        out.child_frame_id = self.output_base_frame
        out.pose.pose.position.x = float(base_translation_std[0])
        out.pose.pose.position.y = float(base_translation_std[1])
        out.pose.pose.position.z = float(base_translation_std[2])
        out.pose.pose.orientation.x = qx
        out.pose.pose.orientation.y = qy
        out.pose.pose.orientation.z = qz
        out.pose.pose.orientation.w = qw
        out.twist = msg.twist
        self.odom_pub.publish(out)

        if self.publish_tf:
            tf_msg = TransformStamped()
            tf_msg.header = out.header
            tf_msg.child_frame_id = self.output_base_frame
            tf_msg.transform.translation.x = out.pose.pose.position.x
            tf_msg.transform.translation.y = out.pose.pose.position.y
            tf_msg.transform.translation.z = out.pose.pose.position.z
            tf_msg.transform.rotation = out.pose.pose.orientation
            self.tf_broadcaster.sendTransform(tf_msg)

    def on_raw_cloud(self, msg: PointCloud2):
        """
        将 Fast-LIO 已配准到 camera_init/raw 世界系的点云转成标准 grounded 世界系。

        这里只保留 x/y/z，open3d_loc 的 ICP 不依赖 intensity/ring/time 字段。
        如果后续需要颜色或强度，可以在这里扩展字段复制逻辑。
        """
        try:
            points = pc2.read_points_numpy(msg, field_names=("x", "y", "z"), skip_nans=True)
        except Exception as exc:
            self.get_logger().warn(f"failed to read raw cloud: {exc}")
            return

        if points.size == 0:
            return

        xyz_raw = np.asarray(points[:, :3], dtype=np.float64)
        xyz_std = (self.raw_to_std @ (xyz_raw - self.initial_body_to_base_translation_raw).T).T

        out = pc2.create_cloud_xyz32(msg.header, xyz_std.astype(np.float32))
        out.header.frame_id = self.odom_frame
        self.cloud_pub.publish(out)


def main(args=None):
    rclpy.init(args=args)
    node = FastLioOpen3dAxisAdapter()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/humanoid_prior_localization_runtime/src/fastlio_open3d_axis_adapter.cpp
/*
 * fastlio_open3d_axis_adapter.cpp
 *
 * 文件作用：
 * 1. 将 Fast-LIO 原始 /odom 与 /fast_lio/cloud_registered 转换为 Open3D 先验地图定位使用的标准轴数据。
 * 2. 发布 /prior_localization/open3d_input_odom，作为 global_localization_node 的里程计初值。
 * 3. 发布 /prior_localization/open3d_input_cloud，作为 global_localization_node 的点云配准输入。
 * 4. 可发布 odom->prior_open3d_base TF，供 prior_map_odom_bridge_cpp 在 odom cache 不足时兜底查询。
 *
 * 上游节点：
 * - fast_lio_node：发布 /odom 和 /fast_lio/cloud_registered。
 *
 * 下游节点：
 * - global_localization_node：使用标准轴 odom/点云做 Open3D 配准。
 * - prior_map_odom_bridge_cpp：使用标准轴 odom cache 对齐定位时间戳。
 *
 * 注意：
 * - 这里只做坐标轴和点云字段转换，不做全局定位和滤波。
 * - initial_body_to_base_translation_raw 必须和 Fast-LIO/body 到导航 base 的外参保持一致。
 */

#include <cmath>
#include <cstddef>
#include <memory>
#include <string>
#include <vector>

#include <Eigen/Core>
#include <Eigen/Geometry>

#include "geometry_msgs/msg/transform_stamped.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/point_cloud2.hpp"
#include "sensor_msgs/point_cloud2_iterator.hpp"
#include "tf2_ros/transform_broadcaster.h"

namespace
{

Eigen::Vector3d parameter_vector3(
  rclcpp::Node & node,
  const std::string & name,
  const std::vector<double> & default_value)
{
  const auto value = node.declare_parameter<std::vector<double>>(name, default_value);
  if (value.size() != 3) {
    RCLCPP_WARN(
      node.get_logger(),
      "parameter %s must contain exactly 3 values; using default [%.3f, %.3f, %.3f]",
      name.c_str(), default_value[0], default_value[1], default_value[2]);
    return Eigen::Vector3d(default_value[0], default_value[1], default_value[2]);
  }
  return Eigen::Vector3d(value[0], value[1], value[2]);
}

}  // namespace

class FastLioOpen3dAxisAdapter : public rclcpp::Node
{
public:
  explicit FastLioOpen3dAxisAdapter(const rclcpp::NodeOptions & options)
  : Node("fastlio_open3d_axis_adapter", options)
  {
    raw_odom_topic_ = declare_parameter<std::string>("raw_odom_topic", "/odom");
    raw_cloud_topic_ = declare_parameter<std::string>("raw_cloud_topic", "/fast_lio/cloud_registered");
    output_odom_topic_ = declare_parameter<std::string>(
      "output_odom_topic", "/prior_localization/open3d_input_odom");
    output_cloud_topic_ = declare_parameter<std::string>(
      "output_cloud_topic", "/prior_localization/open3d_input_cloud");
    odom_frame_ = declare_parameter<std::string>("odom_frame", "odom");
    output_base_frame_ = declare_parameter<std::string>("output_base_frame", "prior_open3d_base");
    publish_tf_ = declare_parameter<bool>("publish_tf", true);

    // Kept for compatibility with older bag-test launch files. The current adapter does not use
    // map_origin_*; the initial_body_to_base_translation_raw parameter defines the cloud origin.
    declare_parameter<double>("map_origin_x", -18.5);
    declare_parameter<double>("map_origin_y", -10.5);
    declare_parameter<double>("map_origin_z", 0.0);

    raw_to_std_ <<
      0.0, 0.0, -1.0,
      1.0, 0.0, 0.0,
      0.0, -1.0, 0.0;

    body_to_base_rotation_raw_ <<
      0.0, 1.0, 0.0,
      0.0, 0.0, -1.0,
      -1.0, 0.0, 0.0;

    initial_body_to_base_translation_raw_ = parameter_vector3(
      *this, "initial_body_to_base_translation_raw", {0.004, 1.215, 0.072});

    rclcpp::QoS input_cloud_qos(rclcpp::KeepLast(3));
    input_cloud_qos.best_effort();
    rclcpp::QoS output_cloud_qos(rclcpp::KeepLast(3));
    output_cloud_qos.reliable();

    odom_pub_ = create_publisher<nav_msgs::msg::Odometry>(output_odom_topic_, 30);
    cloud_pub_ = create_publisher<sensor_msgs::msg::PointCloud2>(output_cloud_topic_, output_cloud_qos);
    tf_broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);

    odom_sub_ = create_subscription<nav_msgs::msg::Odometry>(
      raw_odom_topic_, 50,
      std::bind(&FastLioOpen3dAxisAdapter::onRawOdom, this, std::placeholders::_1));
    cloud_sub_ = create_subscription<sensor_msgs::msg::PointCloud2>(
      raw_cloud_topic_, input_cloud_qos,
      std::bind(&FastLioOpen3dAxisAdapter::onRawCloud, this, std::placeholders::_1));

    RCLCPP_INFO(
      get_logger(),
      "fastlio_open3d_axis_adapter started: %s->%s, %s->%s, tf=%s->%s",
      raw_odom_topic_.c_str(), output_odom_topic_.c_str(),
      raw_cloud_topic_.c_str(), output_cloud_topic_.c_str(),
      odom_frame_.c_str(), output_base_frame_.c_str());
  }

private:
  void onRawOdom(const nav_msgs::msg::Odometry::SharedPtr msg)
  {
    const auto & pose = msg->pose.pose;
    Eigen::Quaterniond raw_quat(
      pose.orientation.w,
      pose.orientation.x,
      pose.orientation.y,
      pose.orientation.z);
    if (raw_quat.norm() == 0.0) {
      raw_quat = Eigen::Quaterniond::Identity();
    } else {
      raw_quat.normalize();
    }

    const Eigen::Matrix3d raw_rotation = raw_quat.toRotationMatrix();
    const Eigen::Vector3d raw_translation(
      pose.position.x,
      pose.position.y,
      pose.position.z);

    const Eigen::Vector3d base_translation_raw =
      raw_translation + raw_rotation * initial_body_to_base_translation_raw_;
    const Eigen::Vector3d base_translation_std =
      raw_to_std_ * (base_translation_raw - initial_body_to_base_translation_raw_);
    const Eigen::Matrix3d base_rotation_std =
      raw_to_std_ * raw_rotation * body_to_base_rotation_raw_;

    Eigen::Quaterniond base_quat(base_rotation_std);
    base_quat.normalize();

    nav_msgs::msg::Odometry out;
    out.header.stamp = msg->header.stamp;
    out.header.frame_id = odom_frame_;
    out.child_frame_id = output_base_frame_;
    out.pose.pose.position.x = base_translation_std.x();
    out.pose.pose.position.y = base_translation_std.y();
    out.pose.pose.position.z = base_translation_std.z();
    out.pose.pose.orientation.x = base_quat.x();
    out.pose.pose.orientation.y = base_quat.y();
    out.pose.pose.orientation.z = base_quat.z();
    out.pose.pose.orientation.w = base_quat.w();
    out.twist = msg->twist;
    odom_pub_->publish(out);

    if (publish_tf_) {
      geometry_msgs::msg::TransformStamped tf_msg;
      tf_msg.header = out.header;
      tf_msg.child_frame_id = output_base_frame_;
      tf_msg.transform.translation.x = out.pose.pose.position.x;
      tf_msg.transform.translation.y = out.pose.pose.position.y;
      tf_msg.transform.translation.z = out.pose.pose.position.z;
      tf_msg.transform.rotation = out.pose.pose.orientation;
      tf_broadcaster_->sendTransform(tf_msg);
    }
  }

  void onRawCloud(const sensor_msgs::msg::PointCloud2::SharedPtr msg)
  {
    std::vector<float> xyz;
    xyz.reserve(static_cast<std::size_t>(msg->width) * static_cast<std::size_t>(msg->height) * 3U);

    try {
      sensor_msgs::PointCloud2ConstIterator<float> iter_x(*msg, "x");
      sensor_msgs::PointCloud2ConstIterator<float> iter_y(*msg, "y");
      sensor_msgs::PointCloud2ConstIterator<float> iter_z(*msg, "z");

      for (; iter_x != iter_x.end(); ++iter_x, ++iter_y, ++iter_z) {
        const float x = *iter_x;
        const float y = *iter_y;
        const float z = *iter_z;
        if (!std::isfinite(x) || !std::isfinite(y) || !std::isfinite(z)) {
          continue;
        }

        xyz.push_back(static_cast<float>(-(static_cast<double>(z) - initial_body_to_base_translation_raw_.z())));
        xyz.push_back(static_cast<float>(static_cast<double>(x) - initial_body_to_base_translation_raw_.x()));
        xyz.push_back(static_cast<float>(-(static_cast<double>(y) - initial_body_to_base_translation_raw_.y())));
      }
    } catch (const std::exception & exc) {
      RCLCPP_WARN_THROTTLE(get_logger(), *get_clock(), 2000, "failed to read raw cloud: %s", exc.what());
      return;
    }

    if (xyz.empty()) {
      return;
    }

    sensor_msgs::msg::PointCloud2 out;
    out.header = msg->header;
    out.header.frame_id = odom_frame_;

    sensor_msgs::PointCloud2Modifier modifier(out);
    modifier.setPointCloud2Fields(
      3,
      "x", 1, sensor_msgs::msg::PointField::FLOAT32,
      "y", 1, sensor_msgs::msg::PointField::FLOAT32,
      "z", 1, sensor_msgs::msg::PointField::FLOAT32);
    modifier.resize(xyz.size() / 3U);
    out.is_dense = true;

    sensor_msgs::PointCloud2Iterator<float> out_x(out, "x");
    sensor_msgs::PointCloud2Iterator<float> out_y(out, "y");
    sensor_msgs::PointCloud2Iterator<float> out_z(out, "z");
    for (std::size_t i = 0; i < xyz.size(); i += 3U, ++out_x, ++out_y, ++out_z) {
      *out_x = xyz[i];
      *out_y = xyz[i + 1U];
      *out_z = xyz[i + 2U];
    }

    cloud_pub_->publish(out);
  }

  std::string raw_odom_topic_;
  std::string raw_cloud_topic_;
  std::string output_odom_topic_;
  std::string output_cloud_topic_;
  std::string odom_frame_;
  std::string output_base_frame_;
  bool publish_tf_{true};

  Eigen::Matrix3d raw_to_std_{Eigen::Matrix3d::Identity()};
  Eigen::Matrix3d body_to_base_rotation_raw_{Eigen::Matrix3d::Identity()};
  Eigen::Vector3d initial_body_to_base_translation_raw_{0.004, 1.215, 0.072};

  rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr odom_pub_;
  rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr cloud_pub_;
  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
  rclcpp::Subscription<sensor_msgs::msg::PointCloud2>::SharedPtr cloud_sub_;
  std::unique_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::NodeOptions options;
  options.automatically_declare_parameters_from_overrides(false);
  rclcpp::spin(std::make_shared<FastLioOpen3dAxisAdapter>(options));
  rclcpp::shutdown();
  return 0;
}
